### ISSUE TITLE
RNNT Decoder outputting blank only

### ISSUE THREAD

**Issue by HendrikLaux** [NVIDIA/NeMo#5946]
I am trying to train a Conformer-Encoder based ASR model on a relatively small and bad quality dataset. So far, the model trains via a CTC loss to about 58% WER and I tried to extend it to use a hybrid CTC/RNNT loss as it can also be done in NeMo. 
Due to the necessity to adapt to an existing codebase, I am not running the provided NeMo scripts/models for training, but rather I create an instance of the required classes for RNNT-models (RNNTLoss, RNNTJoiner, RNNTDecoder) and then include those into my own PyTorch Lightning module. To get all tensors right and fed to the correct modules, I went through the nemo code step by step with a debugger while running the default _speech_to_text_hybrid_rnnt_ctc_bpe.py_ script.  

I feed the zero-padded and tokenized transcripts to the decoder and the output goes to the joiner along with the conformer encoding. I then use the resulting tensors to obtain the RNNT-Loss:
```
decoder, target_length, states = self.decoder(targets=transcript, target_length=transcript_len)
joint = self.joint(encoder_outputs=encoded, decoder_outputs=decoder)
loss_value = self.loss(log_probs=joint, targets=transcript, input_lengths=encoded_len, target_lengths=target_length)
```
(which is basically how it is done in the training step of _nemo/collections/asr/models/rnnt_models.py_ )

So far this works well and I can see the RNNT-Loss dropping (from around 180 to 60 for 'mean_batch' reduction), both when using it alone and when using it together with a CTC loss.
I observed that the performance (in terms of WER) when decoding the CTC output layer's output stays about constant. I am not really sure whether using the RNNT-Loss along with the CTC-Loss will also improve the quality of the CTC decoding, so I decided to try the RNNT decoding. 

Same principle, I instantiated the RNNTDecoding object directly by passing the configuration and the decoder network and joiner network of my model to create a RNNTDecoding object.
However, passing my conformer encoding outputs to _RNNTDecoding.rnnt_decoder_predictions_tensor(...)_  (of the RNNTDecoding instantiated with my joiner and decoder) always results in an empty output/hypothesis with score 0. 
Taking a deeper look into the decoding process (in _ _greedy_decode_blank_as_pad()_ ) shows that in every step the log_probability (logp) of the blank token is extremely high (around 10-12) while lying around -2 .. 2 for all other tokens, which in the end results in an empty hypothesis. This seems especially weird to me since a log_probability is not meant to be positive. 
To get more insight, I ran a pretrained RNNT model (large conformer transducer) and took a look a the same place in the code. While my encoder and decoder outputs look similar (and also f and g) in terms of the value range, logp will always be around -100 .. -20 for this model, but never positive. 
Going into the "joint" step showed that there is some form of log_softmax calculation, but only if the signal is on the cpu. I noticed a similar thing in the RNNT loss already, where (as far as I understood) the log-softmax calculation is handled by numba in case of a GPU tensor.

Formulating an actual question seems hard (and I hope that this is "suited" to be an issue anyway, sorry if it isn't .. ) .. 
But why does the usage of the RNNT decoding fail, although the loss seems to work fine?

Is there any general misconception that I have using the RNNT elements of the library? Are they not meant to be used outside of the NeMo world? 

Why does the logp-signal in   _ _greedy_decode_blank_as_pad_ seems to be log-softmaxed in the pretrained model but this operation is obviously not performed when the class is used in an isolated way? If there is a specific reason, is there a way around it?

**Comment by titu1994** [comment]
This is most definitely suited to be an issue. I think there is a minor bug in hybrid model decoding which will be fixed in the next release (of you can use the r1.16.0 branch for early access once @VahidooX submits it). However that patch is not really going to solve your current issue. 

RNNT is not special or anything, it can most definitely be used outside the NeMo world. But it is a very complex model and one mistake will lead to incorrect results. For example, here is HF script that basically does similar step to your's and removes PTL and adds HF Trainer instead 

https://github.com/sanchit-gandhi/asr-rnnt/tree/main

For RNNT, we don't explicitly compute log_softmax() during training, cuda kernel computes it efficiently. More so, you don't need to apply log softmax in order to compute label via argmax so that step can be skipped - see https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py#L231 and when it's called - https://github.com/NVIDIA/NeMo/blob/main/nemo/collections/asr/parts/submodules/rnnt_greedy_decoding.py#L685

It will call if only for CPU, which actually is also not required but done to maintain consistency with training. What you are seeing are logits, not logprobs, that's why there are positive values. 

From what I can surmise, the main issue is data quality. Loss reducing to 60 is not saying very much cause spec augment artificially increases the loss to a high value as it's "lower bound". For example, without spec augment RNNT loss can and should reach 0 on a toy dataset, but with spec augment can hover around 30-70. 

Question ; how many epochs did training go on for? RNNT / CTC loss plateau for a short period then will steadily decline similar to other losses until it plateaus again. 

What you need to try are two things - switch off spec augment entirely and see if loss stagnates again around a high value.  Second is to enable "log_prediction: true" in config (look up the config it should be at the top) and see if during training the model has started to predict tokens or not. If you used exp manager or wandb, you can refer to your logs for older run. 

What I think is happening is model is simply not learning to predict tokens yet. 

ASR is a two phase alignment learning process - first part is to replace all timesteps with blanks, second part is to replace certain blanks with tokens.

**Comment by titu1994** [comment]
This is pr that fixes the issue, mainly config related change to CTC weight in the hybrid loss - https://github.com/NVIDIA/NeMo/pull/5950

**Comment by VahidooX** [comment]
There was a bug in the hybrid ASR which would hurt the accuracy significantly. Here is the PR to fix the bug in the hybrid ASR model:
https://github.com/NVIDIA/NeMo/pull/5950

You don't need to upgrade your nemo code to use hybrid ASR. You may just copy the new files:
nemo/collections/asr/models/hybrid_rnnt_ctc_models.py
nemo/collections/asr/models/hybrid_rnnt_ctc_bpe_models.py
examples/asr/asr_hybrid_transducer_ctc/speech_to_text_hybrid_rnnt_ctc_bpe.py
examples/asr/conf/conformer/hybrid_transducer_ctc/conformer_hybrid_transducer_ctc_bpe.yaml

**Comment by titu1994** [comment]
RNNT is much more sample efficient than CTC usually. To the point we can use 200 epochs of DON'T ls to converge to a much better results on most datasets than 1000 epochs of CTC training. 

This result is quite odd. I think maybe something we can try as a debugging step is to fine-tune a pretrained RNNT model instead of trying from scratch. 

If regular rnnt also fails to learn then it's not an issue with hybrid model but with the RNNT model itself. One thing to try is reduce the model size to conformer RNNT small or medium to see if those can get RNNT to predict tokens or not. 

RNNT is less data hungry thab CTC, but it does improve more than ctc with sufficient data. The problem seems to be initialization - it is reducing loss but 60 to RNNT loss is not the same as 60 for CTC loss, so maybe it's still not begun training yet v

I would try in order of things - use pretrained encoder, disabling spec aug, reducing model size of above two don't work c

**Comment by HendrikLaux** [comment]
Indeed it seems that the RNNT loss alone doesn't learn as well. Quite stupid of me not to try it in an isolated way first, but probably I was too optimistic for it to directly work .. 

Using RNNT alone gives me the following behavior for various combinations of hyperparameters:
[Training loss curves](https://ibb.co/YfTmd41)
(x-axis is batches and one epoch is around 2.5k batches with BS16)

Here I tried:
- fixed LR (1e-4) + Adam + L-Conformer for encoding (black) - my baseline, which works well for CTC. 
- S/M conformer for the encoder (d_model = 176 or 256), which is dark blue / green.
- Noam Warmup+Annealing instead of fixed LR  (violet)

Further in the red curve I tried to use the frozen RNNTDecoder from the pretrained "stt_en_conformer_transducer_large" together with the tokenizer from this pretrained model. The idea was, that my small dataset may not be able to train the decoder sufficiently with text data, so I use the pretrained one. 

However, it is clearly visible, that:
(1) all the loss curves converge to some value between 6 and 7 pretty quickly (even the experimental one with the pretrained decoder).
(2) My baseline reaches NaN losses at some point (which is the drop to "zero" at around 73k iterations). 

The lower absolute loss values (compared to my previous comment) comes from using the "mean"-reduction now which will divide the individual losses by the target length, but the behaviour is similar for all reduction-variants.

Guess I am doing something massively wrong here. I don't think the problem lies in the rest of the network, because I get pretty decent results when using a CTC loss. And this is basically the exact same network and training loop (except a CTC decoder linear layer on top of the conformer instead of RNNTDecoder and RNNTJoint). 

So I will give your suggestions a try over the weekend and then come back to you, thank you so much for your input!
